Change ShowMergeCommits to HideMergeCommits

gerhardol opened the pull request:
Follow up to #11065 

## Proposed changes

This better reflects the Git defaults and changes the only option ticked by default, that adds filter by selection.
Supposedly, this was a confusion in #11065

Change default for --simplify-merges to false to match defaults. Anyway, the value is not used by default as --full-history is false and --simplify-merges depends on --full-history.

--

This handles a setting with negation in the name, which can be confusing. The cange could be done in the user facing part only.
However, it is normally easier if the code and UI matches. This also remove the special handlig for the setting, negating by default.

## Screenshots <!-- Remove this section if PR does not change UI -->

### Before

![image](https://github.com/gitextensions/gitextensions/assets/6248932/d2b563e2-85e7-45b6-9e3b-f9733330d6f3)

### After

![image](https://github.com/gitextensions/gitextensions/assets/6248932/36cfd23c-d225-445c-947a-2fd249d89844)

## Test methodology <!-- How did you ensure quality? -->

Tests updated

## Merge strategy

I agree that the maintainer squash merge this PR (if the commit message is clear).

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

RussKie commented:
I'm seeing this change as a value-remove not a value-add. The "negative" wording has higher cognitive load, e.g., you can't imaging an _absence_ of something while you can imagine an _existence_ of that. If you may remember the "Configuration" settings page featured negative wording which we inverted in https://github.com/gitextensions/gitextensions/pull/9372.

RussKie commented:
Perhaps, what we should do here is show a "Reset to defaults" button when the filter is changed (hidden by default). E.g., when this is happening:
![image](https://github.com/gitextensions/gitextensions/assets/4403806/1073f4ea-e4d0-4cac-862b-46a7e66ee8db)

This way the user will see that the filter is being applied. What do you think?

gerhardol commented:
The Git default and normally expected is reversed, so a negated interface here should be considered. No checkbox - no filter.

The drop-down menu for filter button already have a reset filter action.
Maybe the icon should be redesigned?
This is a separate (even if related) change for me.

RussKie commented:
> The drop-down menu for filter button already have a reset filter action.

I _think_ the confusion in https://github.com/gitextensions/gitextensions/issues/11065 was partly caused by the lack of visible information that the filter is set - a button to reset could help with awareness...

> Maybe the icon should be redesigned?

No strong opinion (especially given that we lack in graphic design department).

> The Git default and normally expected is reversed, so a negated interface here should be considered. No checkbox - no filter.

I can see that side of the argument too. I'm torn... 
Git isn't know for being very user friendly or user oriented, it's designed by a geek for geeks. The whole purpose of the app is to democratise git and make it more approachable to an average person. So, an inversion of "negative" git settings is in our charter, and is something we've been doing.

gerhardol commented:
Slightly OT - also not this PR

We could stop considering Reflog a filter and remove it from Advanced filter
Maybe move to Commits in View menu. At least one change to reflogs in every release?

With that, Advanced filter (except Full history) is only limiting commits

![image](https://github.com/gitextensions/gitextensions/assets/6248932/d5f1220a-b720-4662-bb9f-c11de0626c64)

RussKie commented:
> Perhaps, what we should do here is show a "Reset to defaults" button when the filter is changed (hidden by default). E.g., when this is happening: ![image](https://user-images.githubusercontent.com/4403806/248453647-1073f4ea-e4d0-4cac-862b-46a7e66ee8db.png)
> 
> This way the user will see that the filter is being applied. What do you think?


https://github.com/gitextensions/gitextensions/assets/4403806/8937c4dd-37fd-4405-bfdc-e4301e11a6cb

What do you think?

gerhardol commented:
> What do you think?

Since all settings are off by default, the need for an reset button is not big. I do not expect users to add many filters, so the main benefit is to indicate that a filter is active. The button is also an indicator, so OK for me.

The reset button does not work right now, no action.

Note: The video must be viewed fullscreen, as the controls hides the button now.

RussKie commented:
My change is a dud for demonstration purposes only. The reset needs to be
wired.

A notable fix - FilterInfo.HasFilter is currently inaccurate.

On Fri, 30 June 2023, 7:29 am Gerhard Olsson, ***@***.***>
wrote:

> What do you think?
>
> Since all settings are off by default, the need for an reset button is not
> big. I do not expect users to add many filters, so the main benefit is to
> indicate that a filter is active. The button is also an indicator, so OK
> for me.
>
> The reset button does not work right now, no action.
>
> Note: The video must be viewed fullscreen, as the controls hides the
> button now.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/11069#issuecomment-1613832612>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABBTEXX2AQVKO3VRBVJIQ5DXNXXVFANCNFSM6AAAAAAZSB2W7E>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

gerhardol commented:
> A notable fix - FilterInfo.HasFilter is currently inaccurate.

HasFilter works. It controls the filter button image.

Reflog is not a filter and full history and simlifymerges are really settings, but View menu is already full. Number of commits is a preference. (All i  recall from top of my head, more settings.)

gerhardol commented:
> What do you think?

I would like to revert this commit.

For instance, the changed `HasFilter()` need to be replaced with a new `IsDefault()` to not set the modified icon when the user sets `Current branch only`.

Suggested to be moved to future PRs:
* Remove refilter from Advanced filter?
* Move `Full history` and `Simplify merges` to View menu, new group `Path filter` or to new page Settings-Detailed-Revision grid (move some other settings like Settings-General-Perfomance-LimitCommits.?
* `Simplify merges` depends on `Full history`
* Reset to default button in Advanced filter form?

RussKie commented:
Sounds good

On Mon, 3 July 2023, 5:40 am Gerhard Olsson, ***@***.***>
wrote:

> What do you think?
>
> I would like to revert this commit.
>
> For instance, the changed HasFilter() need to be replaced with a new
> IsDefault() to not set the modified icon when the user sets Current
> branch only.
>
> Suggested to be moved to future PRs:
>
>    - Remove refilter from Advanced filter?
>    - Move Full history and Simplify merges to View menu, new group Path
>    filter or to new page Settings-Detailed-Revision grid (move some other
>    settings like Settings-General-Perfomance-LimitCommits.?
>    - Simplify merges depends on Full history
>    - Reset to default button in Advanced filter form?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/11069#issuecomment-1616796638>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABBTEXSHB3X5ZRE6NTHIO2DXOHFEXANCNFSM6AAAAAAZSB2W7E>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

mstv reviewed (commented):
I have no strong opinion on having everything as positive "Show xxx" or not - for filters.
"Everything unticked" as default gives a clear UX though.
The negation can be camouflaged by "Hide merge commits" - which I prefer over "No merge commits".
I like "_Show_ only first parent" & "_Show_ merge commits".

gerhardol commented:
>  The negation can be camouflaged by "Hide merge commits" - which I prefer over "No merge commits". 

Change that

> I like "_Show_ only first parent" & "_Show_ merge commits".

Changed to `Show only first parent`

> Show full history

Not changing, this is specific to filtered paths (rewritten history), the option should be somewhere else.
(ended up here as there were no better place in menus, and Settings were considered hiding too much).

mstv reviewed (approved):
LGTM, have not run

RussKie commented:
> Reflog is not a filter and full history and simlifymerges are really settings, but View menu is already full. Number of commits is a preference. (All i recall from top of my head, more settings.)

Technically reflog may not be a filter, but from a user perspective it affects the list of displayed revisions, thus can be viewed as a "filter"-like setting. The same applies to other "_Show Xyz_" settings - those are filter-like too from the user perspective.

When I added the "reset" button (and changed the `HasFilter` condition to include the other settings) I was immediately visually alerted when I enabled reflog and other settings that affected the list of the displayed revisions, and I very much liked that.

gerhardol commented:
> > Reflog is not a filter and full history and simlifymerges are really settings, but View menu is already full. Number of commits is a preference. (All i recall from top of my head, more settings.)
> 
> Technically reflog may not be a filter, but from a user perspective it affects the list of displayed revisions, thus can be viewed as a "filter"-like setting. The same applies to other "_Show Xyz_" settings - those are filter-like too from the user perspective.
> 
> When I added the "reset" button (and changed the `HasFilter` condition to include the other settings) I was immediately visually alerted when I enabled reflog and other settings that affected the list of the displayed revisions, and I very much liked that.

But "Only current branch" is Git default and I do not want a notification when viewing all branches.
Similar, some prefer current-only and do not want a notification for current-only.
I do not want Reflog to be in my filter, the button show that.
Full-history and simply-merges is just a setting that applies for path filters, more a preference.

Reset filters and reset-to-default are separate actions.
